services/postgres/alembic/versions/2020-08-08T205817_remove_state_id_foreign_keys.py

```diff
@@ -0,0 +1,256 @@
+                                FROM new_state
+                                ORDER BY flow_run_id, version DESC, timestamp DESC
+                            ) AS new_latest_state
+                        WHERE
+                            new_latest_state.flow_run_id = flow_run.id
+                            AND (
+                                flow_run.version IS NULL
+                                OR (
+                                    new_latest_state.version >= COALESCE(flow_run.version, -1)
+                                    AND new_latest_state.timestamp > COALESCE(flow_run.state_timestamp, '2000-01-01')));
+                        RETURN NEW;
+                    END;
+                    $$;
+
+
+        CREATE FUNCTION update_task_run_with_latest_state() RETURNS trigger
+            LANGUAGE plpgsql
+            AS $$
+                    BEGIN
+                        UPDATE task_run
+                        SET
+                            version = new_latest_state.version,
+                            -- below are fields that we only include for backwards-compatibility
+                            state=new_latest_state.state,
+                            serialized_state=new_latest_state.serialized_state,
+                            state_start_time=new_latest_state.start_time,
+                            state_timestamp=new_latest_state.timestamp,
+                            state_message=new_latest_state.message,
+                            state_result=new_latest_state.result
+                        FROM
+                            -- only select the latest row
+                            (
+                                SELECT DISTINCT ON (task_run_id)
+                                    id,
+                                    version,
+                                    task_run_id,
+                                    state,
+                                    serialized_state,
+                                    start_time,
+                                    timestamp,
+                                    message,
+                                    result
+                                FROM new_state
+                                ORDER BY task_run_id, version DESC, timestamp DESC
+                            ) AS new_latest_state
+                        WHERE
+                            new_latest_state.task_run_id = task_run.id
+                            AND (
+                                task_run.version IS NULL
+                                OR (
+                                    new_latest_state.version >= COALESCE(task_run.version, -1)
+                                    AND new_latest_state.timestamp > COALESCE(task_run.state_timestamp, '2000-01-01')));
+                        RETURN NEW;
+                    END;
+                    $$;
+
+            CREATE TRIGGER update_flow_run_after_inserting_state
+            AFTER INSERT ON flow_run_state
```

Comment:
Ah - if I'm reading this correctly the trigger only runs after a new state is placed in this table, yea?

Reply:
That's correct, this is the trigger that populates the `flow_run` table with the info from the latest flow run state. It has to be modified here to no longer update the `state_id`